client/core/core.go

```diff
@@ -494,6 +497,86 @@ func (dc *dexConnection) ack(msgID uint64, matchID order.MatchID, signable msgjs
 	return nil
 }
 
+// register submits a new 'register' request to the server. Depending on the
+// server version, either encKey or encKeyLegacy will be set, but not both.
+func (c *Core) register(dc *dexConnection, crypter encrypt.Crypter) (regRes *msgjson.RegisterResult,
+	acctPubB, encKey, encKeyLegacy []byte, paid bool, err error) {
+
+	creds := c.creds()
+
+	// register is called before the *dexConnection is in the Core.conns map,
+	// so no risk of concurrent access.
+	regV2 := dc.acct.dexPubKey != nil
+
+	var accountSuspended bool
+	for {
+		if regV2 && !accountSuspended {
+			encKey, acctPubB, err = dc.acct.setupCryptoV2(creds, crypter)
+			if err != nil {
+				return nil, nil, nil, nil, false, newError(acctKeyErr, "setupCryptoV2 error: %v", err)
+			}
+		} else {
+			encKey = nil
+			// This is an old-style DEX that doesn't provide their pubkey in the
+			// config response. Gotta use a legacy-style key.
+			encKeyLegacy, acctPubB, err = dc.acct.setupCryptoLegacy(crypter)
+			if err != nil {
+				return nil, nil, nil, nil, false, codedError(acctKeyErr, err)
+			}
+		}
+		dexReg := &msgjson.Register{
+			PubKey: acctPubB,
+			Time:   encode.UnixMilliU(time.Now()),
+		}
+		regRes = new(msgjson.RegisterResult)
+		err = dc.signAndRequest(dexReg, msgjson.RegisterRoute, regRes, DefaultResponseTimeout)
+		if err == nil {
+			break
+		}
+		var msgErr *msgjson.Error
+		if errors.As(err, &msgErr) {
+			switch msgErr.Code {
+			case msgjson.AccountExistsError:
+				// Server provides fee coin as the error message.
+				dc.acct.feeCoin, err = hex.DecodeString(msgErr.Message)
+				if err != nil {
+					err = fmt.Errorf("error decoding coin ID from message %q", msgErr.Message)
+				} else {
+					paid = true
+				}
+				return
+			case msgjson.AccountSuspendedError:
+				// Retry with a legacy-style key.
+				c.log.Infof("HD account key for %s was reported as suspended. Generating a legacy-style key instead.", dc.acct.host)
+				accountSuspended = true
+				dc.acct.resetKeys()
+				continue
+			}
+		}
+
+		return nil, nil, nil, nil, false, codedError(registerErr, err)
+	}
+
+	// If we already had a DEX pubkey from the 'config' response, check for
+	// discrepancies.
+	if dc.acct.dexPubKey != nil && !bytes.Equal(dc.acct.dexPubKey.SerializeCompressed(), regRes.DEXPubKey) {
+		return nil, nil, nil, nil, false, fmt.Errorf("different pubkeys reported by dex in 'config' and 'register' responses")
+	}
+
+	// Check the DEX server's signature.
+	msg := regRes.Serialize()
+
+	// If this was a legacy registration, we'll be setting the pubkey here. For
+	// v2 registrations, we're overwriting the existing key, but whatever. We
+	// already checked that they're identical a few lines up.
+	dc.acct.dexPubKey, err = checkSigS256(msg, regRes.DEXPubKey, regRes.Sig)
+	if err != nil {
+		return nil, nil, nil, nil, false, newError(signatureErr, "%s pubkey error: %v", dc.acct.host, err)
+	}
```

Comment:
Relating to my comment about putting DEXPubKey in a 0.2.2 patch, this could go away anyway if we do that.  A patch release is not asking a lot for operators.  It can be applied well in advance of 0.3 too.